File: tests/case_study/Fortran/speedtest_smooth_index.f90
PROGRAM jacobi
  USE ISO_FORTRAN_ENV
  IMPLICIT NONE
  INTEGER, PARAMETER :: NX=10000, NY=10000, NITS=200
  REAL(REAL64), DIMENSION(:,:), ALLOCATABLE,TARGET :: data, data2
  INTEGER :: i, j, its, start,end
  REAL :: count_rate
  ALLOCATE(data(0:NX+1,0:NY+1), data2(0:NX+1,0:NY+1))
  data=0
  data(0,:)=1
  data(NX+1,:)=1
  data(:,0)=1
  data(:,NY+1)=1
  CALL SYSTEM_CLOCK(start,count_rate)
  !$OMP PARALLEL
  DO its=1,NITS
    DO j=1,NY
      !$OMP DO
      DO i=1,NX
        data2(i,j)=0.25_REAL64 * (data(i-1,j)+data(i+1,j)+data(i,j-1)+data(i,j+1))
      ENDDO
    ENDDO
    data(1:NX,1:NY)=data2(1:NX,1:NY)
  ENDDO
  !$OMP END PARALLEL
  CALL SYSTEM_CLOCK(end)
  PRINT *, 'Time for main loop is ', REAL(end-start)/count_rate
  PRINT '(F24.17)', MAXVAL(data(1:NX,1:NY))
  PRINT '(F24.17)', MINVAL(data(1:NX,1:NY))
  PRINT '(F24.17)', SUM(data(1:NX,1:NY))/REAL(SIZE(data(1:NX,1:NY),KIND=INT64),REAL64)

END PROGRAM jacobi

File: tests/case_study/C++/speedtest_smooth_index.cpp

#include <iomanip>
#include "far.h"
#include "timer.h"
#define NX 10000
#define NY 10000
#define ITS 200

using namespace far;

int main(){
	Array<double,2> data(Range(0,NX+1),Range(0,NY+1)), data2;
	data=0;
	data2.mold(data);
	//Set the boundaries on data
	data(0,Range(0,NY+1))=1;
	data(NX+1,Range(0,NY+1))=1;
	data(Range(0,NX+1),0)=1;
	data(Range(0,NX+1),NY+1)=1;


	timer t;
	t.begin("main loop");
#pragma omp parallel
	{
		for (int its = 0; its<ITS;++its){
#ifndef FAR_USE_C_INDEX
			for (int j=1;j<=NY;++j){
				#pragma omp for
				for (int i=1;i<=NX;++i){
					data2(i,j) = (data(i-1,j) + data(i+1,j) + data(i,j-1) + data(i,j+1)) * 0.25;
				}
			}
#else
      for (int i=1;i<=NX;++i){
        #pragma omp for
        for (int j=1;j<=NY;++j){
          data2(i,j) = (data(i-1,j) + data(i+1,j) + data(i,j-1) + data(i,j+1)) * 0.25;
        }
      }
#endif
			//Copy back
      beginWorkshare();
			data(Range(1,NX),Range(1,NY))=data2(Range(1,NX),Range(1,NY));
      endWorkshare();
#pragma omp barrier
		}
	}
	t.end();
	std::cout << std::setprecision(17);
	std::cout << maxval(data(Range(1,NX),Range(1,NY))) << "\n";
	std::cout << minval(data(Range(1,NX),Range(1,NY))) << "\n";
	std::cout << sum(data(Range(1,NX),Range(1,NY)))/float(data(Range(1,NX),Range(1,NY)).getSize()) << "\n";
}
